src/util/Timer.cpp

```diff
@@ -718,4 +724,97 @@ VirtualTimer::async_wait(std::function<void()> const& onSuccess,
         mEvents.push_back(ve);
     }
 }
+
+TimeDebtTracker::TimeDebtTracker(std::string const& name, size_t jubileeLimit)
+    : mName(name), mJubileeLimit(jubileeLimit)
+{
+}
+
+VirtualClock::time_point
+TimeDebtTracker::computeYieldTime(VirtualClock& clock,
+                                  std::chrono::milliseconds intendedDuration)
+{
+    ++mRequestCount;
+    if (mRequestCount >= mJubileeLimit)
+    {
+        mRequestCount = 0;
+        if (mCurrentDebt.count() != 0)
+        {
+            LOG(TRACE) << "TimeDebtTracker " << mName
+                       << " declaring jubilee on " << mCurrentDebt.count()
+                       << "ms of time-debt";
+        }
+        mCurrentDebt = std::chrono::milliseconds(0);
+    }
+
+    VirtualClock::time_point now = clock.now();
+    if (mCurrentDebt.count() > 0)
+    {
+        // Caller is in debt, we'll reduce their request and the corresponding
+        // debt by some amount.
+        std::chrono::milliseconds repayment =
+            std::min(mCurrentDebt, intendedDuration);
+        mCurrentDebt -= repayment;
+        intendedDuration -= repayment;
+    }
+    return now + intendedDuration;
+}
+
+void
+TimeDebtTracker::recordDifferenceFromIntendedYieldTime(
+    VirtualClock& clock, VirtualClock::time_point intendedYieldTime)
+{
+    // Caller was _supposed_ to yield at intendedYieldTime but it actually
+    // yielded right now. Check to see if they ran a time deficit.
+    std::chrono::milliseconds deficit =
+        std::chrono::duration_cast<std::chrono::milliseconds>(
+            clock.now() - intendedYieldTime);
+
+    // We only record "positive" deficits: if work wound up finishing early we
+    // do not record "negative" surpluses to boost future budgets.
+    if (deficit.count() > 0)
+    {
+        mCurrentDebt += deficit;
+        LOG(TRACE) << "TimeDebtTracker " << mName << " noting "
+                   << deficit.count() << "ms of new time-debt, totalling "
+                   << mCurrentDebt.count() << "ms";
+    }
+}
+
+YieldTimer::YieldTimer(VirtualClock& clock, TimeDebtTracker& tbt,
+                       std::chrono::milliseconds yieldAfterDuration,
+                       size_t yieldAfterIteration)
+    : mClock(clock)
+    , mDebtTracker(tbt)
+    , mYieldTime(tbt.computeYieldTime(clock, yieldAfterDuration))
+    , mIterationsRemaining(yieldAfterIteration)
+{
+}
+
+bool
+YieldTimer::shouldKeepGoing()
+{
+    // To make it easier to read meaning of loop headers.
+    return !shouldYield();
+}
+
+bool
+YieldTimer::shouldYield()
+{
+    if (mIterationsRemaining == 0)
+    {
+        mDebtTracker.recordDifferenceFromIntendedYieldTime(mClock, mYieldTime);
```

Comment:
I did this initially but the dtor of a YieldTimer runs some number of steps _after_ the loop exits -- whenever the enclosing scope exits -- so this is not strictly correct. It might be quite a bit incorrect, like if there's an inner transaction committing in there. The timing will be wrong.